### ISSUE TITLE
Feature request G'MIC Python binding <-> PIL input/output missing 8-bits pixel value interleaved codec

### ISSUE THREAD

**Issue by myselfhimself** [python-pillow/Pillow#4954]
### What did you do?
**Could you help me in evaluating if I need to implement a new codec or not in order to simplify WIP [G'MIC](gmic.eu/) <-> PIL Python conversion functions ?**

I am trying to make a PIL.Image object from a GmicImage object in the most optimized way using [gmic-py](https://github.com/dtschump/gmic-py), a C/Python binding for G'MIC.
The goal is to implement `GmicImage.from_PIL()` and `GmicImage.to_PIL()` conversion methods inside the gmic-py Python binding, so that PIL users can pass in image buffers from and to G'MIC as part of their pipeline, without needing to output images first as files, and reopening them in each library/tool.

G'MIC has a single way of storing pixel values through its GmicImage object's `_data` `bytes` attribute: 4 bytes per pixel value (ie. float32 on all OSes), interleaved. So an RGB pixel in G'MIC takes 3 floats = 3 x 4 x 8 = 96bits, all interleaved at image level, no stride.

I have been considering to use the PIL.Image.frombytes() constructor for creating a PIL Image from a G'MIC Image buffer, but cannot find the relevant raw mode.

Tentative code (for now in pure Python POC) is the to_PIL() method below, along with step 3.:
```python
import PIL.Image
import gmic # one must pip install gmic==2.9.1-alpha1
import numpy
%matplotlib inline

def to_PIL(gmicimage):
  # I failed to find a proper raw decoder for 32-bits RGB deinterleaved
  # too lazy to implement a custom decoder for now
  return PIL.Image.frombytes('RGB', (gmicimage._width, gmicimage._height), gmicimage._data, 'raw', 'RGBX') # per https://github.com/python-pillow/Pillow/blob/c841501c3416bf245b5688dd33890641ef0f5dd7/src/libImaging/Unpack.c#L1342

# Color ramp generating command from https://gmic.eu/gallery/colors.shtml since G'MIC samples downloads are blocked by GColab
GMIC_GENERATE_COLORS_IMAGE_COMMAND = "(0,360;0,360^0,0;1,1^1,1;1,1) resize 400,400,1,3,3 hsv2rgb"

# 1. Generate and image and store it to file
l = []
gmic.run(GMIC_GENERATE_COLORS_IMAGE_COMMAND + " output colors.png", l)

# 2. Load the file-output image from PIL and display for visual comparison
im=PIL.Image.open('colors.png')
print(im.width, im.height, im.mode, im.format, type(im))
display(im)

# 3. Make G'MIC Image into a PIL Image through Numpy and display
l_as_PIL = to_PIL(l[0])
print(l_as_PIL.width, l_as_PIL.height, l_as_PIL.mode, l_as_PIL.format, type(l_as_PIL))
display(l_as_PIL)
```

Related Google Colab script; to fiddle with: https://colab.research.google.com/drive/1Ox4vigcEHgEqUjUGFyHh7gl76CYkolt0?usp=sharing

### What did you expect to happen?
I expected a relevant raw mode codec to exist, but have not found or not understood which one to use. I wish to be lazy and avoid coding a new codec myself for now.

### What actually happened?
My image is not properly decoded and does not look like a color ramp at all.

### What are your OS, Python and Pillow versions?

* OS: Linux
* Python: 3.6 probably (Google Colab's version)
* Pillow: most recent stable from pypi.org - 7.2.0 I think

Related issue: https://github.com/myselfhimself/gmic-py/issues/59

**Comment by wiredfool** [comment]
The core issue is that Pillow doesn't support multiband multibyte images. So, a single band float is ok, and a multiband 8 bit per channel is ok, but a 4 channel float isn't supported.

**Comment by myselfhimself** [comment]
Thank you having taken the time to think and reply!

Indeed I was thinking of providing a partially degraded/destructive I/O support, with data mutation if needed. Native RGB (ie. if I understand well 3 bands x 32 bits float interleaved) GmicImage -> PIL.Image conversion by introducing a GmicImage.to_PIL() method with float->int (0:255) casting with some already available raw codec in some PIL.Image class factory method/constructor.

For now, performance is not a priority and I prefer skipping writing an additional codec.. so I have found a workaround in both input and output direction using numpy as an intermediary tool - which is lossless until PIL ingesting takes place - and PIL.Image fromarray() and toarray(), methods:

```python
import gmic
import PIL.Image
import numpy

def from_PIL(PILimage):
  gmicimage = gmic.GmicImage.from_numpy_array(numpy.array(PILimage), deinterleave=True)
  return gmicimage

def to_PIL(gmicimage, astype=numpy.uint8, squeeze=True, mode='RGB'):
  arr = gmicimage.to_numpy_array(astype=numpy.uint8)
  return PIL.Image.fromarray(arr.squeeze() if squeeze else arr, mode)
```

Looking at the very last return statement, it seems that PIL.Image.fromarray() does not like ingesting non-squeezed shapes if a provided z axis provided at index 2 (eg. shapes like (y,x,z=1,c) are forbidden, in favor of (y,x,c)). I was unsure whether PIL allows for some volumetric dimension to exist, a bit like how TIFF multi-image or scikitimage multi-planes. I might be mistaken in my interpretation... for now 2D image I/O, with or without a z=1 shape axes is OK.

**Comment by myselfhimself** [comment]
This can be closed. Have a nice life :)
